### PR TITLE
Tweak audit types docstring

### PR DESCRIPTION
**Pull request by tengqm**
#### What type of PR is this?

/kind cleanup
/kind documentation

#### What this PR does / why we need it:

This PR tweaks the doc strings for audit.v1 types. Primary issues fixed include:

- Avoid using field names as seen by developers. The user for the generated config API mostly don't care about the implementation details.
- Avoid using special characters like '*' that may confuse the config API generator (including a markdown parser) which believes a star means bold text.
- Provide more useful information for type alias such as 'Level', 'Stage'.

#### Which issue(s) this PR fixes:

N/A

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by dims** [comment]
@tengqm looks like we need to regenerate stuff (see verify CI job failure)

**Comment by tengqm** [comment]
@dims Thanks for the review. I'm again stuck by the 'hack/update-generated-protobuf.sh` script, because I'm behind a damn wall:

```
$ hack/update-generated-protobuf.sh
+++ [0605 22:21:19] Verifying Prerequisites....
+++ [0605 22:21:19] Using docker on macOS
+++ [0605 22:21:23] Building Docker image kube-build:build-4500b067c5-5-v1.25.0-go1.18.2-bullseye.0
+++ Docker build command failed for kube-build:build-4500b067c5-5-v1.25.0-go1.18.2-bullseye.0

#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 1.92kB 0.0s done
#1 DONE 0.1s

#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for registry.k8s.io/build-image/kube-cross:v1.25.0-go1.18.2-bullseye.0
#3 ERROR: failed to do request: Head "https://k8s.gcr.io/v2/build-image/kube-cross/manifests/v1.25.0-go1.18.2-bullseye.0": Service Unavailable
------
 > [internal] load metadata for registry.k8s.io/build-image/kube-cross:v1.25.0-go1.18.2-bullseye.0:
------
error: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to do request: Head "https://k8s.gcr.io/v2/build-image/kube-cross/manifests/v1.25.0-go1.18.2-bullseye.0": Service Unavailable

To retry manually, run:

DOCKER_CLI_EXPERIMENTAL=enabled docker buildx build --load -t kube-build:build-4500b067c5-5-v1.25.0-go1.18.2-bullseye.0 --pull=false --build-arg=KUBE_CROSS_IMAGE=registry.k8s.io/build-image/kube-cross --build-arg=KUBE_CROSS_VERSION=v1.25.0-go1.18.2-bullseye.0 /Users/tengqm/k8s/src/k8s.io/kubernetes/_output/images/kube-build:build-4500b067c5-5-v1.25.0-go1.18.2-bullseye.0

!!! [0605 22:22:40] Call tree:
!!! [0605 22:22:40]  1: hack/../build/run.sh:31 kube::build::build_image(...)
```

**Comment by dims** [comment]
@tengqm there's a `diff` of what got changed in the build log:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/110354/pull-kubernetes-verify/1532371487770546176/build-log.txt

If you apply that, that should make it go green.

**Comment by tengqm** [comment]
> @tengqm there's a `diff` of what got changed in the build log: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/110354/pull-kubernetes-verify/1532371487770546176/build-log.txt
> 
> If you apply that, that should make it go green.

That is a great tip!!!

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go [hunk 1]
```diff
@@ -72,17 +72,27 @@ const (
...
 	// Stage of the request handling when this event instance was generated.
+	// The value can be one of:
+	// - 'RequestReceived': The stage for events generated as soon as the audit handler receives
+	//   the request, and before it is delegated down the handler chain.
+	// - 'ResponseStarted': The stage for events generated once the response headers are sent,
+	//   but before the response body is sent. This stage is only generated for long-running requests
+	//   (e.g. watch).
+	// - 'ResponseComplete': The stage for events generated once the response body has been completed,
+	//   and no more bytes will be sent.
+	// - 'Panic': The stage for events generated when a panic occurred.
```
Comment (roycaihw) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:90]: there was a concern about using markdown in API doc: https://github.com/kubernetes/kubernetes/pull/109259#issuecomment-1117099014. I want to make sure it's on your radar

**Review by roycaihw — COMMENTED** [review]
/lgtm
/approve

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go [hunk 2]
```diff
@@ -72,17 +72,27 @@ const (
...
 	Stage Stage `json:"stage" protobuf:"bytes,3,opt,name=stage,casttype=Stage"`
 
-	// RequestURI is the request URI as sent by the client to a server.
+	// The request URI as sent by the client to a server.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:93]: isn't `requestURI is the request ...` the expected style, i.e. the JSON field name? Has the suggested style changed?

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go (hunk 1 above)
Comment (tengqm) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:90]: I'm aware of that comment, but I don't agree with it. For OpenAPI 2.0, it is said that "for description, ([GFM syntax](https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown) can be used for rich text representation)". For OpenAPI 3.0, it is said that "Throughout the specification description fields are noted as supporting CommonMark markdown formatting.".

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go (hunk 2 above)
Comment (tengqm) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:93]: The "expected style" is a little bit vague. We can say `requestURI is the request ...` if all developers are comfortable with it. If we are more aggressive at it, we should say ```The `requestURL` is the request...```. The second style is much clearer when the field name is a regular English word, e.g. ``` `request` is the request received ...```. However, I'm not sure our developers are comfortable with adding extra backticks. The "style" adopted by this PR is actually a compromise between developer habits and API consumers.

Comment (sftim) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:93]: To be honest, most API documentation comments use UpperCamelCase names. Making a change to this feels like it'd be a bigger change than this PR.

I'd like to find a way to have the end-users see relevant API docs (and for the OpenAPI to use the right capitalization, too) - but that does feel like a KEP-sized work package.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go [hunk 3]
```diff
@@ -72,17 +72,27 @@ const (
...
 	RequestURI string `json:"requestURI" protobuf:"bytes,4,opt,name=requestURI"`
-	// Verb is the kubernetes verb associated with the request.
+	// The kubernetes verb associated with the request.
```
Comment (sftim) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:95]:
```suggestion
	// The Kubernetes API verb associated with the request.
```

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go [hunk 4]
```diff
@@ -219,88 +243,104 @@ type PolicyRule struct {
...
 	// +optional
 	ResourceNames []string `json:"resourceNames,omitempty" protobuf:"bytes,3,rep,name=resourceNames"`
 }
 
 // ObjectReference contains enough information to let you inspect or modify the referred object.
 type ObjectReference struct {
+	// The kind of the referred object.
 	// +optional
 	Resource string `json:"resource,omitempty" protobuf:"bytes,1,opt,name=resource"`
+	// The namespace of the referred object.
```
Comment (sftim) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:324]:
```suggestion
	// The namespace of the referred object. Only set for requests made to
	// namespace-scoped APIs.
```
?

**Review by sftim — CHANGES_REQUESTED** [review]
Some thoughts, happy to see this merge without the changes I've suggested.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go (hunk 3 above)
Comment (tengqm) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:95]: done

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go (hunk 1 above)
Comment (sftim) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:90]: I agree with **OpenAPI** using Markdown. I'm not convinced that as a project we have committed to treating _code comments_ as Markdown.

**This looks OK to me**: the single quotes are appropriate here for a sort of halfway house between a commitment to Markdown and not.

I would like to see some specially marked doc comments treated (and linted) as Markdown; maybe one day I'd have the cycles for a KEP, and the project could have the cycles to adopt it. Plenty of of work there, too.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go [hunk 5]
```diff
@@ -219,88 +243,104 @@ type PolicyRule struct {
...
-	// 'pods' matches pods.
-	// 'pods/log' matches the log subresource of pods.
-	// '*' matches all resources and their subresources.
-	// 'pods/*' matches all subresources of pods.
-	// '*/scale' matches all scale subresources.
+	// - `pods` matches pods.
+	// - `pods/log` matches the log subresource of pods.
+	// - `*` matches all resources and their subresources.
+	// - `pods/*` matches all subresources of pods.
+	// - `*/scale` matches all scale subresources.
```
Comment (sftim) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:304]: I wonder if we should single quotes here. We don't have a strong convention of being able to treat code comments as valid Markdown and we should be wary of adopting that by stealth.

**Review by sftim — COMMENTED** [review]
All the changes that change from an UpperCamelCase name to a part of speech such as “The” or “This” feel to me like we're ducking a bigger question about turning code comments into documentation.

I woudn't do it. This is a personal opinion (I'm not speaking for SIG Docs), but: informal fixes that infer a _de facto_ convention tend to narrow our options for future work.

I'd rather accept that the generated docs aren't quite right and look for project-wide consensus on an approach that we like and that we feel we can staff appropriately. We've have 25 minor versions with imperfect API docs. I'd rather have more of those than find the project painting itself into code maintenance corner without realizing.

Long term, I think we all want the same thing: good API docs, readable code, and scope to have automation (not humans) point out things that everyone would agree are a mistake.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go (hunk 5 above)
Comment (tengqm) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go:304]: @sftim You may want to read the OpenAPI spec if you don't have an idea about the assumption of having markdown style description for types and fields. The fact is that strings like these are making the API docs confusing, and the confusion is not limited to the configuration APIs but also all API references. I can give you a lot of other examples if needed.

**Comment by tengqm** [comment]
> All the changes that change from an UpperCamelCase name to a part of speech such as “The” or “This” feel to me like we're ducking a bigger question about turning code comments into documentation.
> 

In an ideal world, we may want to change all cases where a field name is spelt as how a Go developer sees it to how a API user sees it. However, this is impractical. We can argue on that for months or years. The rephrasing of "The", "This" is a compromise so that the developers know what it is and the API user know what it is. It hurts nobody and it is a quick and clean fix.

**Pull request closed without merging**